subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.execution;
+
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.tasks.TaskExecuter;
+import org.gradle.api.internal.tasks.TaskExecutionContext;
+import org.gradle.api.internal.tasks.TaskStateInternal;
+import org.gradle.initialization.BuildCancellationToken;
+
+public class CancellableTaskExecuter implements TaskExecuter {
+    private final TaskExecuter delegate;
+    private final BuildCancellationToken buildCancellationToken;
+
+    public CancellableTaskExecuter(TaskExecuter delegate, BuildCancellationToken buildCancellationToken) {
+        this.delegate = delegate;
+        this.buildCancellationToken = buildCancellationToken;
+    }
+
+    @Override
+    public void execute(final TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
```

Comment:
❌ Not sure if we have test cases for this already:

- cancelled tasks don't lead to weird errors on the console
- outputs of cancelled tasks are not snapshotted/cached, the task is rerun the next time

Reply:
Added the test cases here: https://github.com/gradle/gradle/blob/0211eb748345bc88222b985530c1383ddb479232/subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy#L64

Reply:
Missing test case for cachable tasks.

Reply:
Added test cases with `--build-cache` enabled and assertions of no build status being printed on client console.